src/machine/aot/mod.rs

```diff
@@ -0,0 +1,673 @@
+const MAXIMUM_SECTIONS: usize = 1024;
+const MAXIMUM_DUMMY_SECTIONS: usize = 64;
+
+const ADDRESS_WRITE_ONLY_FLAG: u64 = 0x8000_0000_0000_0000;
+const ADDRESS_LABEL_FLAG: u64 = 0x4000_0000_0000_0000;
+const MAXIMUM_ENCODED_ADDRESS: u64 = 0x8000_0000;
+
+#[derive(Debug, Clone)]
+pub enum Write {
+    Memory {
+        address: Value,
+        size: u8,
+        value: Value,
+    },
+    Register {
+        index: usize,
+        value: Value,
+    },
+    Pc {
+        value: Value,
+    },
+    Ecall,
+    Ebreak,
+}
+
+fn init_registers() -> [Value; 32] {
+    [
+        Value::Imm(0),
+        Value::Register(1),
+        Value::Register(2),
+        Value::Register(3),
+        Value::Register(4),
+        Value::Register(5),
+        Value::Register(6),
+        Value::Register(7),
+        Value::Register(8),
+        Value::Register(9),
+        Value::Register(10),
+        Value::Register(11),
+        Value::Register(12),
+        Value::Register(13),
+        Value::Register(14),
+        Value::Register(15),
+        Value::Register(16),
+        Value::Register(17),
+        Value::Register(18),
+        Value::Register(19),
+        Value::Register(20),
+        Value::Register(21),
+        Value::Register(22),
+        Value::Register(23),
+        Value::Register(24),
+        Value::Register(25),
+        Value::Register(26),
+        Value::Register(27),
+        Value::Register(28),
+        Value::Register(29),
+        Value::Register(30),
+        Value::Register(31),
+    ]
+}
+
+struct LabelGatheringMachine {
+    registers: [Value; 32],
+    pc: Value,
+    labels_to_test: Vec<u64>,
+
+    // A memory segment which contains code loaded from ELF
+    memory: FlatMemory<u64>,
+    labels: HashSet<u64>,
+    sections: Vec<(u64, u64)>,
+    dummy_sections: HashMap<u64, u64>,
+}
+
+impl LabelGatheringMachine {
+    pub fn load(program: &Bytes) -> Result<Self, Error> {
+        let elf = Elf::parse(&program).unwrap();
+        if elf.section_headers.len() > MAXIMUM_SECTIONS {
+            return Err(Error::LimitReached);
+        }
+        let mut sections: Vec<(u64, u64)> = elf
+            .section_headers
+            .iter()
+            .filter_map(|section_header| {
+                if section_header.sh_flags & u64::from(SHF_EXECINSTR) != 0 {
+                    Some((
+                        section_header.sh_addr,
+                        section_header.sh_addr + section_header.sh_size,
+                    ))
+                } else {
+                    None
+                }
+            })
+            .rev()
+            .collect();
+        // Test there's no empty section
+        if sections.iter().any(|(s, e)| s >= e) {
+            return Err(Error::OutOfBound);
+        }
+        // Test no section overlaps with one another. We first sort section
+        // list by start, then we test if each end is equal or less than
+        // the next start.
+        sections.sort_by_key(|section| section.0);
```

Comment:
Just some quick question, not code change suggestions.

1. Can code jump between sections?
2. If so, will sorting sections help improve the performance a bit? Since the RISC-V code may already have optimized the code to increase the chance to do sequential read on jumping. If we keep the same sequence in the x86 ASM, we may benefit from this optimization.

Reply:
1. Yes code can jump freely between sections
2. It's hard to say if this is gonna bring any improvement, normal GCC outputs actually only have one text section for all the code, and code reorganization happens within this single section.